lib/spack/spack/ci.py

```diff
@@ -828,63 +887,87 @@ def generate_gitlab_ci_yaml(env, print_summary, output_file,
     else:
         tty.warn('Unable to populate buildgroup without CDash credentials')
 
-    if final_job_config and not is_pr_pipeline:
-        # Add an extra, final job to regenerate the index
-        final_stage = 'stage-rebuild-index'
-        final_job = {
-            'stage': final_stage,
-            'script': 'spack buildcache update-index --keys -d {0}'.format(
-                mirror_urls[0]),
-            'tags': final_job_config['tags'],
-            'when': 'always'
-        }
-        if 'image' in final_job_config:
-            final_job['image'] = final_job_config['image']
-        if before_script:
-            final_job['before_script'] = before_script
-        if after_script:
-            final_job['after_script'] = after_script
-        output_object['rebuild-index'] = final_job
-        stage_names.append(final_stage)
-
-    output_object['stages'] = stage_names
-
-    # Capture the version of spack used to generate the pipeline, transform it
-    # into a value that can be passed to "git checkout", and save it in a
-    # global yaml variable
-    spack_version = spack.main.get_version()
-    version_to_clone = None
-    v_match = re.match(r"^\d+\.\d+\.\d+$", spack_version)
-    if v_match:
-        version_to_clone = 'v{0}'.format(v_match.group(0))
-    else:
-        v_match = re.match(r"^[^-]+-[^-]+-([a-f\d]+)$", spack_version)
+    nonbuild_job_config = None
+    if 'nonbuild-job-attributes' in gitlab_ci:
+        nonbuild_job_config = gitlab_ci['nonbuild-job-attributes']
+
+    attrib_list = ['image', 'tags', 'variables']
+
+    if job_id > 0:
+        if enable_rebuild_index and not is_pr_pipeline:
+            # Add an extra, final job to regenerate the index
+            final_stage = 'stage-rebuild-index'
+            final_job = {
+                'stage': final_stage,
+                'script': 'spack buildcache update-index --keys -d {0}'.format(
+                    mirror_urls[0]),
+                'when': 'always'
+            }
+            if nonbuild_job_config:
+                copy_attributes(attrib_list, nonbuild_job_config, final_job)
+            if before_script:
+                final_job['before_script'] = before_script
```

Comment:
I think this is a bug (also right below, where we set the `after_script` on the `final_job`).  Here we have been using whichever (if any) `before_script` and `after_script` were associated with the final spec in the matrix, and applying them to the `rebuild-index` job.  It works in some cases, but could easily fail to produce a `rebuild-index` job that does what you want.  Not to mention it's not at all clear where these values (`before_script` and `after_script`) are coming from for the `rebuild-index` job.

To fix it, I could add `before_script` and `after_script` to the `nonbuild-job-attributes` section and the populate the `rebuild-index` job with those.  However, we decided the `noop` job (produced in the case that DAG pruning results in an otherwise empty pipeline)  would also pull its `tags` and `image` or `variables` from the `nonbuild-job-attributes` section.  Requiring these two jobs to use the same runner attributes is starting to feel a little wonky.  The `noop` job could easily run on a tiny instance, does not need spack, `boto3`, or even python for that matter.  It just needs a `script` with a single line: `echo "Nothing to do"`.  On the other hand the `rebuild-index` job does need python, spack, and likely `boto3`, probably needs a `before_script` to clone spack and source it, as well as an `after_script` to clean things up, etc. 

For this reason, I'm now wondering if it makes more sense to require the user to define a `rebuild-index` job explicitly in their `.gitlab-ci.yml`, if they need it.  It suddenly feels weird to have the user supply us with the `tags`, `image` + `entrypoint` or `variables`, `before_script`, `after_script`, just so we can then provide the line that does `spack buildcache update-index ...` with the right arguments.  We would obviously document this (why do we not rebuild the index at the end of each job, why do we need an updated index, etc...) and provide an example of a `rebuild-index` job and how it fits into a spack environment pipeline configuration.

The `noop` job could still use the `nonbuild-job-attributes` to get its `tags` and `image` or `variables`, as I have so far found no way we could let the user define this job themselves, or get around the need for it.

Reply:
The only thing I *don't* like about removing the `rebuild-index` key from the `gitlab-ci` schema is that if we no longer generate that job, but instead require the user to configure it in their `.gitlab-ci.yml`, then I don't know of a way to make running it conditional on there being at least one spec rebuilt.  That was easy when the job was generated with the pipeline.

Reply:
Considering it can take such a long time to rebuild an index, especially when it lives in an S3 bucket, I'm reconsidering removing the generation of a `rebuild-index` job.  Our cloud PR/`develop` pipelines must have this job at the end if any specs were rebuilt and pushed to the mirror, so eliminating generation of the `rebuild-index` job means we'll have to add that job to our `.gitlab-ci.yml` and it will run every time.  If it takes 10 - 30 minutes to generate the index, that becomes the ~minimum~ wait time for the gitlab ci PR check to finish, assuming everything else takes zero time.  

I have a change I'm going to push on this PR today which takes advantage of buildcache indices to reduce drastically the time to generate a PR with DAG pruning, and with both of these changes, we could see some pipelines turning around in 5 minutes when nothing needs to be done.  

However, an additional consideration is the api we expose in the schema for assigning attributes to generated jobs.  Currently this PR leaves us with only one set of attributes for all non-build jobs.  So far, those could include 1) clean up per-pipeline mirror, 2) regenerate buildcache index, and 3) no-op job needed for otherwise empty DAG-pruned pipelines).  This means that ~each~ such job will need the union of all the attributes needed for ~all~ such jobs.  As a specific example, this means the no-op job which simply echoes "nothing to do" will also need spack cloned and sourced, `boto3` installed, etc... because those are required by the job which regenerates the buildcache index.  While this might be tolerable, if not desirable, it seems like it could cause confusion to folks try to reason about how all this works.

Some alternatives are to have a specific schema providing attributes for each type of non-build job we know about, or to group some job types them somehow in order to achieve  smaller set of schema items.  For example, we could have one set of attributes for the rebuild index job, and then all the other non-build attributes could share a second set of attributes.

Reply:
I decided to generate a `rebuild-index` job, so it can be skipped when DAG pruning results in an empty pipeline.  To fix the bug mentioned at the top of this discussion, I look for the optional `service-job-attributes` and apply those to the `rebuild-index` job.